### PR TITLE
Port //daml-lf/data to Scala 2.13

### PR DESCRIPTION
**Pull request by cocreature**
Apologies for the giant PR, I couldn’t find an intermediate state that compiles unfortunately. This was fairly painful since the whole point of this library is to define collection types and these are more or less incompatible between 2.12 and 2.13 so there is a decent amount of duplication. I’ve tried to limit it as much as possible without compromising on efficiency.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/ImmArray.scala [hunk 1]
```diff
@@ -150,20 +143,20 @@ final class ImmArray[+A] private (
 
   /** O(n) */
   def map[B](f: A => B): ImmArray[B] = {
-    val newArray: mutable.ArraySeq[B] = new mutable.ArraySeq(length)
+    val newArray: Array[Any] = new Array(length)
```
Comment (cocreature) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/ImmArray.scala:146]: I’m not super happy about the `Array[Any]` part but that seemed like the most reasonable option to make this work. Open to other suggestions.

**Comment by cocreature** [comment]
Benchmark results from `CollectAuthority`:

Before
```
Benchmark                                                             (dar)             (scenario)  Mode  Cnt   Score   Error  Units
CollectAuthority.bench  //daml-lf/scenario-interpreter/CollectAuthority.dar  CollectAuthority:test  avgt  200  21.508 ± 0.093  ms/op
```
After
```
Benchmark                                                             (dar)             (scenario)  Mode  Cnt   Score   Error  Units
CollectAuthority.bench  //daml-lf/scenario-interpreter/CollectAuthority.dar  CollectAuthority:test  avgt  200  21.253 ± 0.087  ms/op
```

Lower is better but either way, seems like it’s just noise.

**Comment by S11001001** [comment]
f30e15eb7be68e490f048e2365d86536a8626dc6, which I pushed on top of this branch, factors the ImmArraySeq commonalities, relegating the Scala-version-specific files to divergent implementations.

**Comment by cocreature** [comment]
Nice I like it! Do you think the other duplication is also worth reducing or should we keep that at least for this PR?

**Comment by S11001001** [comment]
> Do you think the other duplication is also worth reducing

Yes,

* [ ] InsertOrdSet
* [ ] InsertOrdMap

deserve the same treatment.

**Review by S11001001 — APPROVED** [review]

**Comment by cocreature** [comment]
Thanks for the changes @S11001001!

**Review by remyhaemmerle-da — APPROVED** [review]

**Pull request merged**